AssertionError in pdfminer.six

We have another AssertionError from pdfminer that we need to fix in `pdftotree`.
```
$ pdftotree -vv LITES00424-1.pdf 
[INFO] pdftotree.core - Digitized PDF detected, building tree structure...
INFO:pdftotree.core:Digitized PDF detected, building tree structure...
[ERROR] pdftotree.TreeExtract - (316.495, 316.495, 50)
Traceback (most recent call last):
  File "/home/lwhsiao/repos/pdftotree/pdftotree/TreeExtract.py", line 148, in get_candidates_alignments
    nodes, features = parse_layout(elems, font_stat)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 56, in parse_layout
    boxes_segments, boxes_curves, boxes_figures, page_width, combine)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 67, in cluster_vertically_aligned_boxes
    plane.extend(boxes)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 341, in extend
    self.add(obj)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 346, in add
    for k in self._getrange((obj.x0, obj.y0, obj.x1, obj.y1)):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 334, in _getrange
    for x in drange(x0, x1, self.gridsize):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 171, in drange
    assert v0 < v1, str((v0, v1, d))
AssertionError: (316.495, 316.495, 50)
ERROR:pdftotree.TreeExtract:(316.495, 316.495, 50)
Traceback (most recent call last):
  File "/home/lwhsiao/repos/pdftotree/pdftotree/TreeExtract.py", line 148, in get_candidates_alignments
    nodes, features = parse_layout(elems, font_stat)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 56, in parse_layout
    boxes_segments, boxes_curves, boxes_figures, page_width, combine)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 67, in cluster_vertically_aligned_boxes
    plane.extend(boxes)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 341, in extend
    self.add(obj)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 346, in add
    for k in self._getrange((obj.x0, obj.y0, obj.x1, obj.y1)):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 334, in _getrange
    for x in drange(x0, x1, self.gridsize):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 171, in drange
    assert v0 < v1, str((v0, v1, d))
AssertionError: (316.495, 316.495, 50)
```

Another example output:

```
$ pdftotree -vv LITES00424-1.pdf 
[INFO] pdftotree.core - Digitized PDF detected, building tree structure...
INFO:pdftotree.core:Digitized PDF detected, building tree structure...
[ERROR] pdftotree.TreeExtract - (316.495, 316.495, 50)
Traceback (most recent call last):
  File "/home/lwhsiao/repos/pdftotree/pdftotree/TreeExtract.py", line 148, in get_candidates_alignments
    nodes, features = parse_layout(elems, font_stat)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 56, in parse_layout
    boxes_segments, boxes_curves, boxes_figures, page_width, combine)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 67, in cluster_vertically_aligned_boxes
    plane.extend(boxes)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 341, in extend
    self.add(obj)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 346, in add
    for k in self._getrange((obj.x0, obj.y0, obj.x1, obj.y1)):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 334, in _getrange
    for x in drange(x0, x1, self.gridsize):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 171, in drange
    assert v0 < v1, str((v0, v1, d))
AssertionError: (316.495, 316.495, 50)
ERROR:pdftotree.TreeExtract:(316.495, 316.495, 50)
Traceback (most recent call last):
  File "/home/lwhsiao/repos/pdftotree/pdftotree/TreeExtract.py", line 148, in get_candidates_alignments
    nodes, features = parse_layout(elems, font_stat)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 56, in parse_layout
    boxes_segments, boxes_curves, boxes_figures, page_width, combine)
  File "/home/lwhsiao/repos/pdftotree/pdftotree/utils/pdf/pdf_parsers.py", line 67, in cluster_vertically_aligned_boxes
    plane.extend(boxes)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 341, in extend
    self.add(obj)
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 346, in add
    for k in self._getrange((obj.x0, obj.y0, obj.x1, obj.y1)):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 334, in _getrange
    for x in drange(x0, x1, self.gridsize):
  File "/home/lwhsiao/repos/pdftotree/.venv/lib/python3.5/site-packages/pdfminer/utils.py", line 171, in drange
    assert v0 < v1, str((v0, v1, d))
AssertionError: (316.495, 316.495, 50)
```

--- a/bin/pdftotree
+++ b/bin/pdftotree
@@ -44,6 +44,12 @@ if __name__ == '__main__':
         action='store_true',
         help="Whether to output visualization images for the tree")
     parser.add_argument(
+        '-d',
+        '--dry-run',
+        dest='dry_run',
+        action='store_true',
+        help="Run pdftotree, but do not save any output or print to console.")
+    parser.add_argument(
         '-v',
         '--verbose',
         dest='verbose',
@@ -77,11 +83,16 @@ if __name__ == '__main__':
 
     log.addHandler(ch)
 
+    if args.dry_run:
+        print("This is just a dry run. No HTML will be output.")
+        args.output = None
+
     # Call the main routine
     result = pdftotree.parse(args.pdf_file, args.output, args.model_path,
                              args.favor_figures, args.visualize)
 
     if result:
-        print(result)
+        if not args.dry_run:
+            print(result)
     else:
         print("HTML output to {}".format(args.output))

--- a/bin/pdftotree
+++ b/bin/pdftotree
@@ -73,14 +73,12 @@ if __name__ == '__main__':
 
     # Configure logging for this application
     log = logging.getLogger('pdftotree')
-    log.setLevel(log_level)
-
+    log.propagate = 0  # prevent propagation to the root logger
     ch = logging.StreamHandler()
+    log.setLevel(log_level)
     ch.setLevel(log_level)
-
     formatter = logging.Formatter("[%(levelname)s] %(name)s - %(message)s")
     ch.setFormatter(formatter)
-
     log.addHandler(ch)
 
     if args.dry_run:

--- a/pdftotree/TreeExtract.py
+++ b/pdftotree/TreeExtract.py
@@ -117,10 +117,11 @@ class TreeExtractor(object):
             page_num, elems)
         # print "Page Num: ", page_num, "Line bboxes: ", len(lines_bboxes), ", Alignment bboxes: ", len(alignments_bboxes)
         # alignment_features += get_alignment_features(lines_bboxes, elems, font_stat)
-        boxes = alignments_bboxes  # + lines_bboxes
+        boxes = alignments_bboxes
         if len(boxes) == 0:
             log.info("No boxes were found on page {}.".format(page_num))
             return [], []
+
         lines_features = get_lines_features(boxes, elems)
         features = np.concatenate((np.array(alignment_features),
                                    np.array(lines_features)), axis=1)
@@ -201,10 +202,6 @@ class TreeExtractor(object):
 
             # TODO: We need to detect columns and sort acccordingly.
             boxes.sort(key=cmp_to_key(column_order))
-
-            #  from pprint import pprint
-            #  pprint(boxes, width=120)
-            #  import pdb; pdb.set_trace()
 
             for box in boxes:
                 if(box[0] == "table"):

--- a/pdftotree/TreeExtract.py
+++ b/pdftotree/TreeExtract.py
@@ -119,6 +119,7 @@ class TreeExtractor(object):
         # alignment_features += get_alignment_features(lines_bboxes, elems, font_stat)
         boxes = alignments_bboxes  # + lines_bboxes
         if len(boxes) == 0:
+            log.info("No boxes were found on page {}.".format(page_num))
             return [], []
         lines_features = get_lines_features(boxes, elems)
         features = np.concatenate((np.array(alignment_features),

--- a/pdftotree/TreeExtract.py
+++ b/pdftotree/TreeExtract.py
@@ -119,7 +119,7 @@ class TreeExtractor(object):
         # alignment_features += get_alignment_features(lines_bboxes, elems, font_stat)
         boxes = alignments_bboxes
         if len(boxes) == 0:
-            log.info("No boxes were found on page {}.".format(page_num))
+            self.log.info("No boxes were found on page {}.".format(page_num))
             return [], []
 
         lines_features = get_lines_features(boxes, elems)

--- a/pdftotree/core.py
+++ b/pdftotree/core.py
@@ -46,9 +46,7 @@ def parse(pdf_file,
           model_path=None,
           favor_figures=True,
           visualize=False):
-
     log = logging.getLogger(__name__)
-
     model = None
     if (model_path is not None):
         model = load_model(model_path)

--- a/pdftotree/core.py
+++ b/pdftotree/core.py
@@ -58,7 +58,7 @@ def parse(pdf_file,
         pdf_tree = extractor.get_tree_structure(model, favor_figures)
         log.info("Tree structure built, creating html...")
         pdf_html = extractor.get_html_tree()
-        log.info("HTML created, outputting...")
+        log.info("HTML created.")
         pdf_filename = os.path.basename(pdf_file)
         # Check html_path exists, create if not
         pdf_html = re.sub(r'[\x00-\x1F]+', '', pdf_html)

--- a/pdftotree/utils/pdf/layout_utils.py
+++ b/pdftotree/utils/pdf/layout_utils.py
@@ -10,7 +10,6 @@ from itertools import chain
 from pdfminer.layout import LTTextLine, LTChar, LTAnno, LTCurve, LTComponent, LTLine
 from pdftotree.utils.pdf.vector_utils import *
 
-log = logging.getLogger(__name__)
 
 def traverse_layout(root, callback):
     '''
@@ -23,7 +22,7 @@ def traverse_layout(root, callback):
             traverse_layout(child, callback)
 
 
-def get_near_items(tree,tree_key):
+def get_near_items(tree, tree_key):
     '''
     Check both possible neighbors for key
     in a binary tree
@@ -37,7 +36,8 @@ def get_near_items(tree,tree_key):
     except KeyError:
         pass
 
-def align_add(tree, key, item, align_thres = 2.0):
+
+def align_add(tree, key, item, align_thres=2.0):
     '''
     Adding the item object to a binary tree with the given
     key while allow for small key differences
@@ -52,15 +52,18 @@ def align_add(tree, key, item, align_thres = 2.0):
     tree[key] = [item]
 
 
-right_wall = lambda m: (m.x1,m.y0,m.x1,m.y1)
-left_wall = lambda m: (m.x0,m.y0,m.x0,m.y1)
-top_wall = lambda m: (m.x0,m.y0,m.x1,m.y0)
-bot_wall = lambda m: (m.x0,m.y1,m.x1,m.y1)
+right_wall = lambda m: (m.x1, m.y0, m.x1, m.y1)
+left_wall = lambda m: (m.x0, m.y0, m.x0, m.y1)
+top_wall = lambda m: (m.x0, m.y0, m.x1, m.y0)
+bot_wall = lambda m: (m.x0, m.y1, m.x1, m.y1)
+
+
 def vlines_between(plane, prev, m):
     if not prev or not m: return []
     if prev.xc > m.xc: prev, m = m, prev
     query = (prev.xc, prev.yc, m.xc, prev.yc)
     return [l for l in plane.find(query) if l.x1 - l.x0 < 0.1]
+
 
 def hlines_between(plane, prev, m):
     if not prev or not m: return []
@@ -69,7 +72,7 @@ def hlines_between(plane, prev, m):
     return [l for l in plane.find(query) if l.y1 - l.y0 < 0.1]
 
 
-def is_same_row(m1,m2):
+def is_same_row(m1, m2):
     # Corner case for row separation
     #------
     #-prev- ------
@@ -77,11 +80,15 @@ def is_same_row(m1,m2):
     #       ------
     return m1 and m2 and m2.yc > m1.y0 and m2.yc < m1.y1
 
+
 def is_vline(l):
     return l.x1 - l.x0 < 0.1
 
+
 def is_hline(l):
     return l.y1 - l.y0 < 0.1
+
+
 #
 # def align_add_to_tree(tree, key, item, close_enough_func):
 #     '''
@@ -102,7 +109,8 @@ def is_hline(l):
 #         tree[key] = [item]
 #
 
-def collect_table_content(table_bboxes,elems):
+
+def collect_table_content(table_bboxes, elems):
     '''
     Returns a list of elements that are contained inside
     the corresponding supplied bbox.
@@ -121,12 +129,12 @@ def collect_table_content(table_bboxes,elems):
         # and we can avoid checking all tables for elems inside
         # Elements only need to intersect the bbox for table as some
         # formatting of fonts may result in slightly out of bbox text
-        if prev_bbox is not None and intersect(prev_bbox,c.bbox):
+        if prev_bbox is not None and intersect(prev_bbox, c.bbox):
             prev_content.append(c)
             continue
         # Search the rest of the tables for membership when done with
         # the current one
-        for table_id,table_bbox in enumerate(table_bboxes):
+        for table_id, table_bbox in enumerate(table_bboxes):
             if intersect(table_bbox, c.bbox):
                 prev_bbox = table_bbox
                 prev_content = table_contents[table_id]
@@ -134,8 +142,10 @@ def collect_table_content(table_bboxes,elems):
                 break
     return table_contents
 
+
 _bbox = namedtuple('_bbox', ['bbox'])
 _inf_bbox = _bbox([float('inf')] * 4)
+
 
 def _gaps_from(intervals):
     '''
@@ -147,7 +157,8 @@ def _gaps_from(intervals):
     gaps = [b[0] - a[1] for a, b in sliding_window]
     return gaps
 
-def project_onto(objs, axis, min_gap_size = 4.0):
+
+def project_onto(objs, axis, min_gap_size=4.0):
     '''
     Projects object bboxes onto the axis and return the
     unioned intervals and groups of objects in intervals.
@@ -155,10 +166,10 @@ def project_onto(objs, axis, min_gap_size = 4.0):
     if axis == 'x': axis = 0
     if axis == 'y': axis = 1
     axis_end = axis + 2
-    if axis == 0: # if projecting onto X axis
-        objs.sort(key = lambda o:o.x0)
+    if axis == 0:  # if projecting onto X axis
+        objs.sort(key=lambda o: o.x0)
     else:
-        objs.sort(key = lambda o:o.y0)
+        objs.sort(key=lambda o: o.y0)
 
     intervals = []
     groups = []
@@ -168,7 +179,7 @@ def project_onto(objs, axis, min_gap_size = 4.0):
     end = objs[0].bbox[axis_end]
 
     # Use _inf_bbox to trigger the last interval divide
-    for o_i, o in enumerate(chain(objs,[_inf_bbox])):
+    for o_i, o in enumerate(chain(objs, [_inf_bbox])):
 
         # Get current interval
         o_start = o.bbox[axis]
@@ -193,6 +204,7 @@ def project_onto(objs, axis, min_gap_size = 4.0):
         # else do nothing
     return intervals, groups
 
+
 def recursive_xy_divide(elems, avg_font_size):
     '''
     Recursively group/divide the document by white stripes
@@ -201,14 +213,16 @@ def recursive_xy_divide(elems, avg_font_size):
     avg_font_size: the minimum gap size between elements below
     which we consider interval continuous.
     '''
+    log = logging.getLogger(__name__)
     log.info(avg_font_size)
     objects = list(elems.mentions)
     objects.extend(elems.segments)
     bboxes = []
+
     # A tree that is a list of its children
     # bboxes can be recursively reconstructed from
     # the leaves
-    def divide(objs, bbox, h_split = True, is_single = False):
+    def divide(objs, bbox, h_split=True, is_single=False):
         '''
         Recursive wrapper for splitting a list of objects
         with bounding boxes.
@@ -243,8 +257,7 @@ def recursive_xy_divide(elems, avg_font_size):
 
     full_page_bbox = (0, 0, elems.layout.width, elems.layout.height)
     # Filter out invalid objects
-    objects = [o for o in objects if inside(full_page_bbox,o.bbox)]
+    objects = [o for o in objects if inside(full_page_bbox, o.bbox)]
     log.info('avg_font_size for dividing', avg_font_size)
     tree = divide(objects, full_page_bbox) if objects else []
     return bboxes, tree
-

--- a/pdftotree/utils/pdf/pdf_parsers.py
+++ b/pdftotree/utils/pdf/pdf_parsers.py
@@ -62,7 +62,9 @@ def cluster_vertically_aligned_boxes(boxes, page_bbox, avg_font_pts, width,
                                      boxes_figures, page_width, combine):
     # Too many "." in the Table of Content pages
     if (len(boxes) == 0 or len(boxes) > 3500):
+        log.error("Too many '.' in the Table of Content pages.")
         return []
+
     plane = Plane(page_bbox)
     plane.extend(boxes)
 
@@ -667,11 +669,7 @@ def parse_tree_structure(elems, font_stat, page_num, ref_page_seen, tables,
     width = get_page_width(
         mentions + boxes_segments + boxes_figures + boxes_curves)
 
-    try:
-        char_width = get_char_width(mentions)
-    except Exception as e:
-        log.warning("Unable to get char_width. Defaulting to char_width = 2.")
-        char_width = 2
+    char_width = get_char_width(mentions)
 
     grid_size = avg_font_pts / 2.0
 
@@ -1156,17 +1154,21 @@ def get_most_common_font_pts(mentions, font_stat):
     '''
     font_stat: Counter object of font sizes
     '''
-    # default min font size of 1 pt in case no font present
-    most_common_font_size = font_stat.most_common(1)[0][0]
-    # Corner case when no text on page
-    if not most_common_font_size: return 2.0
-    count = 0.01  # avoid division by zero
-    height_sum = 0.02  # default to pts 2.0
-    for m in mentions:
-        if m.font_size == most_common_font_size:
-            height_sum += m.height
-            count += 1
-    return height_sum / count
+    try:
+        # default min font size of 1 pt in case no font present
+        most_common_font_size = font_stat.most_common(1)[0][0]
+
+        count = 0.01  # avoid division by zero
+        height_sum = 0.02  # default to pts 2.0
+        for m in mentions:
+            if m.font_size == most_common_font_size:
+                height_sum += m.height
+                count += 1
+        return height_sum / count
+
+    except IndexError as ie:
+        log.info("No text found on page. Default most_common_font_pts to 2.0")
+        return 2.0
 
 
 def get_page_width(boxes):
@@ -1185,4 +1187,8 @@ def get_char_width(boxes):
     for i, b in enumerate(boxes):
         box_len_sum = box_len_sum + b.bbox[2] - b.bbox[0]
         num_char_sum = num_char_sum + len(b.get_text())
-    return box_len_sum / num_char_sum
+    try:
+        return box_len_sum / num_char_sum
+    except ZeroDivisionError as ze:
+        log.warning("No text found. Defaulting to char_width = 2.0.")
+        return 2.0

--- a/pdftotree/utils/pdf/pdf_parsers.py
+++ b/pdftotree/utils/pdf/pdf_parsers.py
@@ -60,6 +60,13 @@ def parse_layout(elems, font_stat, combine=False):
 def cluster_vertically_aligned_boxes(boxes, page_bbox, avg_font_pts, width,
                                      char_width, boxes_segments, boxes_curves,
                                      boxes_figures, page_width, combine):
+    # Filter out boxes with zero width or height
+    filtered_boxes = []
+    for bbox in boxes:
+        if (bbox.x1 - bbox.x0 > 0 and bbox.y1 - bbox.y0 > 0):
+            filtered_boxes.append(bbox)
+    boxes = filtered_boxes
+
     # Too many "." in the Table of Content pages
     if len(boxes) == 0:
         log.warning("No boxes were found to cluster.")
@@ -715,7 +722,7 @@ def parse_tree_structure(elems, font_stat, page_num, ref_page_seen, tables,
     else:
         tables_page = tables
 
-    ##Eliminate tables from these boxes
+    ## Eliminate tables from these boxes
     boxes = []
     for idx1, box in enumerate(mentions):
         intersect = False
@@ -765,6 +772,13 @@ def parse_tree_structure(elems, font_stat, page_num, ref_page_seen, tables,
 def extract_text_candidates(boxes, page_bbox, avg_font_pts, width, char_width,
                             page_num, ref_page_seen, boxes_figures, page_width,
                             page_height):
+    # Filter out boxes with zero width or height
+    filtered_boxes = []
+    for bbox in boxes:
+        if (bbox.x1 - bbox.x0 > 0 and bbox.y1 - bbox.y0 > 0):
+            filtered_boxes.append(bbox)
+    boxes = filtered_boxes
+
     #Too many "." in the Table of Content pages - ignore because it takes a lot of time
     if (len(boxes) == 0 or len(boxes) > 3500):
         return {}, False
@@ -1097,6 +1111,13 @@ def extract_text_candidates(boxes, page_bbox, avg_font_pts, width, char_width,
 
 def get_figures(boxes, page_bbox, page_num, boxes_figures, page_width,
                 page_height):
+
+    # Filter out boxes with zero width or height
+    filtered_boxes = []
+    for bbox in boxes:
+        if (bbox.x1 - bbox.x0 > 0 and bbox.y1 - bbox.y0 > 0):
+            filtered_boxes.append(bbox)
+    boxes = filtered_boxes
 
     if len(boxes) == 0:
         log.warning(

--- a/pdftotree/utils/pdf/pdf_parsers.py
+++ b/pdftotree/utils/pdf/pdf_parsers.py
@@ -61,9 +61,12 @@ def cluster_vertically_aligned_boxes(boxes, page_bbox, avg_font_pts, width,
                                      char_width, boxes_segments, boxes_curves,
                                      boxes_figures, page_width, combine):
     # Too many "." in the Table of Content pages
-    if (len(boxes) == 0 or len(boxes) > 3500):
-        log.error("Too many '.' in the Table of Content pages.")
-        return []
+    if len(boxes) == 0:
+        log.warning("No boxes were found to cluster.")
+        return [], []
+    elif len(boxes) > 3500:
+        log.warning("Too many '.' in the Table of Content pages?")
+        return [], []
 
     plane = Plane(page_bbox)
     plane.extend(boxes)
@@ -1094,6 +1097,12 @@ def extract_text_candidates(boxes, page_bbox, avg_font_pts, width, char_width,
 
 def get_figures(boxes, page_bbox, page_num, boxes_figures, page_width,
                 page_height):
+
+    if len(boxes) == 0:
+        log.warning(
+            "No boxes to get figures from on page {}.".format(page_num))
+        return []
+
     plane = Plane(page_bbox)
     plane.extend(boxes)
 

--- a/pdftotree/utils/pdf/pdf_parsers.py
+++ b/pdftotree/utils/pdf/pdf_parsers.py
@@ -19,8 +19,6 @@ from pdfminer.utils import Plane
 from pdftotree.utils.pdf.layout_utils import *
 from pdftotree.utils.pdf.node import Node
 
-log = logging.getLogger(__name__)
-
 
 def parse_layout(elems, font_stat, combine=False):
     '''
@@ -60,6 +58,7 @@ def parse_layout(elems, font_stat, combine=False):
 def cluster_vertically_aligned_boxes(boxes, page_bbox, avg_font_pts, width,
                                      char_width, boxes_segments, boxes_curves,
                                      boxes_figures, page_width, combine):
+    log = logging.getLogger(__name__)
     # Filter out boxes with zero width or height
     filtered_boxes = []
     for bbox in boxes:
@@ -772,6 +771,7 @@ def parse_tree_structure(elems, font_stat, page_num, ref_page_seen, tables,
 def extract_text_candidates(boxes, page_bbox, avg_font_pts, width, char_width,
                             page_num, ref_page_seen, boxes_figures, page_width,
                             page_height):
+    log = logging.getLogger(__name__)
     # Filter out boxes with zero width or height
     filtered_boxes = []
     for bbox in boxes:
@@ -1111,7 +1111,7 @@ def extract_text_candidates(boxes, page_bbox, avg_font_pts, width, char_width,
 
 def get_figures(boxes, page_bbox, page_num, boxes_figures, page_width,
                 page_height):
-
+    log = logging.getLogger(__name__)
     # Filter out boxes with zero width or height
     filtered_boxes = []
     for bbox in boxes:
@@ -1184,6 +1184,7 @@ def get_most_common_font_pts(mentions, font_stat):
     '''
     font_stat: Counter object of font sizes
     '''
+    log = logging.getLogger(__name__)
     try:
         # default min font size of 1 pt in case no font present
         most_common_font_size = font_stat.most_common(1)[0][0]
@@ -1212,6 +1213,7 @@ def get_page_width(boxes):
 
 
 def get_char_width(boxes):
+    log = logging.getLogger(__name__)
     box_len_sum = 0
     num_char_sum = 0
     for i, b in enumerate(boxes):

--- a/pdftotree/utils/pdf/pdf_utils.py
+++ b/pdftotree/utils/pdf/pdf_utils.py
@@ -25,8 +25,6 @@ from pdftotree.utils.pdf.vector_utils import *
 # Compact wrapper representation for the pdf
 PDFElems = namedtuple('PDFElems', ['mentions', 'segments', 'curves', 'figures', 'layout', 'chars'])
 
-log = logging.getLogger(__name__)
-
 
 class CustomPDFPageAggregator(PDFPageAggregator):
     '''
@@ -116,6 +114,7 @@ def analyze_pages(file_name, char_margin=1.0):
             try:
                 interpreter.process_page(page)
             except OverflowError as oe:
+                log = logging.getLogger(__name__)
                 log.exception("{}, skipiping page {} of {}".format(oe, page_num, file_name))
                 continue
             layout = device.get_result()

